src/transformation/visitors/switch.ts

```diff
@@ -1,61 +1,136 @@
 import { FunctionVisitor } from "../context";
-import { unsupportedForTarget } from "../utils/diagnostics";
 import { performHoisting, popScope, pushScope, ScopeType } from "../utils/scope";
 
-export const transformSwitchStatement: FunctionVisitor<ts.SwitchStatement> = (statement, context) => {
-    if (context.luaTarget === LuaTarget.Universal || context.luaTarget === LuaTarget.Lua51) {
-        context.diagnostics.push(unsupportedForTarget(statement, "Switch statements", LuaTarget.Lua51));
+const containsBreakOrReturn = (statements: ts.Node[]): boolean => {
+    for (const s of statements) {
+        if (ts.isBreakStatement(s) || ts.isReturnStatement(s)) {
+            return true;
+        } else if (!ts.isBlock(s)) {
+            // Can only ensure a break scoped in a block is deterministic
+            continue;
+        } else if (containsBreakOrReturn(s.getChildren())) {
+            return true;
+        }
     }
 
+    return false;
+};
+
+export const transformSwitchStatement: FunctionVisitor<ts.SwitchStatement> = (statement, context) => {
     const scope = pushScope(context, ScopeType.Switch);
 
-    // Give the switch a unique name to prevent nested switches from acting up.
+    // Give the switch and condition accumulator a unique name to prevent nested switches from acting up.
     const switchName = `____switch${scope.id}`;
+    const conditionName = `____cond${scope.id}`;
     const switchVariable = lua.createIdentifier(switchName);
+    const conditionVariable = lua.createIdentifier(conditionName);
 
+    // If the switch only has a default clause, wrap it in a single do.
+    // Otherwise, we need to generate a set of if statements to emulate the switch.
     let statements: lua.Statement[] = [];
-
-    // Starting from the back, concatenating ifs into one big if/elseif statement
-    const concatenatedIf = statement.caseBlock.clauses.reduceRight((previousCondition, clause, index) => {
-        if (ts.isDefaultClause(clause)) {
-            // Skip default clause here (needs to be included to ensure index lines up with index later)
-            return previousCondition;
+    const clauses = statement.caseBlock.clauses;
+    if (clauses.length === 1 && ts.isDefaultClause(clauses[0])) {
+        const defaultClause = clauses[0].statements;
+        if (defaultClause.length) {
+            statements.push(lua.createDoStatement(context.transformStatements(defaultClause)));
         }
+    } else {
+        // Build up the condition for each if statement
+        // Fallthrough is handled by accepting the last condition as an additional or clause
+        // Default is the not of all known case expressions
+        let previousClause: ts.CaseOrDefaultClause;
+        let condition: lua.Expression | undefined;
+        let isInitialCondition = true;
+        statement.caseBlock.clauses.forEach(clause => {
+            if (!condition || (previousClause && containsBreakOrReturn([...previousClause.statements]))) {
```

Comment:
I'm confused about the logic here. If the previous clause contained a break/return, the condition is simply assigned, but it will still be wrapped in an `or` down below. Why not do that here like clauses that don't follow a break?

Reply:
Could use a refactor now that all tests are passing. This is just churn on solving the problem.
You sure you don't want to get in on this... ;)

Reply:
Fixed in 8d1f74a105e1994e5ee6a7df0a2851d217fae06b